### DISCUSSION TITLE
Option to `set-language` in selected block

### DISCUSSION THREAD

**Discussion by koalazub** [helix-editor/helix#8230] [Ideas]
Would the option to select a block of text and set it as a different language be on the cards? It's highly beneficial for languages like `go` that utilise `HTML` templates. I'm finding that there's a clash when I switch to `HTML` and then back to `go` when I'm doing some work. 

This could also become a standard expectation, as I had assumed it followed the same behaviour as `:reflow`. If you think I should file this as an enhancement instead of an idea, I'll do so when I can.
Thanks for reading!

**Comment by the-mikedavis** [comment]
It's possible to do this with tree-sitter injections https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection but that typically needs some kind of marker like code fences in markdown. If HTML is recognizable within Go like other languages are in markdown, that could be added to `runtime/queries/go/injections.scm`. 

I expect that the ability to set injections manually would take much larger changes though - I'm not sure it's possible to add that without some really large changes in core.

**Reply by koalazub** [reply]
Code fence could be triggered via backticks in go files when parsing the template. But it would be nice to be able to inject wherever the code block is selected:

```go
import "html/template"
...
t, err := template.New("foo").Parse(`{{define "T"}}Hello, {{.}}!{{end}}`)
err = t.ExecuteTemplate(out, "T", "<script>alert('you have been pwned')</script>")
```
[Excerpt from here](https://pkg.go.dev/html/template#hdr-Introduction)

But as you already pointed out, this definitely seems like it'd be a very large feature implementation.

I could also see this spilling off into intelligently discerning things like SQL code within string literals. 

[JetBrains](https://www.jetbrains.com/help/idea/using-language-injections.html#configure-injection-rules) would be the gold standard of such a feature(I'm just mentioning, not looking at parity). 

The main takes:

- manually inject a language without the need of using code fencing
- Is it worth considering such a feature

Is this worth having a larger conversation? Or would you rather I just pocket this?

**Reply by the-mikedavis** [reply]
I'll put this on the back burner in my head - I can see value in the feature but I'm not sure it's worth the large change I expect that it needs.

I think it would be ideal if we could solve the use-cases where you would want to use this feature with regular tree-sitter injections. For example we have some Rust injections that highlight SQL strings passed to the `sqlx` library as SQL: https://github.com/helix-editor/helix/blob/1c88432efc3724f60b27d580b8b490040a3048f6/runtime/queries/rust/injections.scm#L28-L53

**Reply by koalazub** [reply]
Alrighty so I want to bubble up this discussion with a new problem that I'm running into and that's implementing graphql template into typescript. 

So we've now got two points to make for resolving a similar issue:
- implementing HTML templates in `go` code
- implementing graphql templates in `tsx` code

I'm not sure if you've had to deal with this, but these are moments that I wish I could just inject a language for a set number of lines so that I could invoke the LSP for that language within a scope. 

Do you think this warrants further investigation or maybe even raises the priority for such a feature?

**Reply by the-mikedavis** [reply]
LSP can't work on part of a file, the spec assumes that you're working on entire files and those files exist on disk. The most we could do with a command like this is temporarily enable highlighting within the selected block

**Reply by koalazub** [reply]
Seems like it wasn't made to work like that by design according to the threads I've read. 

Here is what I found: https://github.com/microsoft/language-server-protocol/issues/636

So I guess you might be able to spawn some sort of adjacent server to keep track of the language within a space if you so choose. This is according to the conversation made in the link above. There are a few addendums and a bit of progress on how they've handled such a situation. One of which is terraform's pr listed on the bottom of that same link.

I guess that can make this task even larger in a sense, I imagine
